plugins/usertiming.js

```diff
@@ -33,7 +33,7 @@
 		 * @returns {string} compressed user timing data
 		 */
 		getUserTiming: function() {
-			var timings, res, now = BOOMR.now();
+			var timings, res, now = performance.now();
```

Comment:
unfortunately this is not safe to use if the site uses a UserTiming shim that does not also shim `performance.now()`

Reply:
Yeah...I see now. I was looking at https://github.com/SOASTA/boomerang/blob/master/plugins/usertiming.js#L117, but it does not shim `performance.now()`.

Reply:
@bluesmoon Any chance you can explain [this comment](https://github.com/SOASTA/boomerang/blob/master/boomerang.js#L1190-L1192)? Should `BOOMR.now()` actually be producing the correct timestamp?

Reply:
well, `BOOMR.now()` is supposed to return a unix timestamp using the best available timing mechanism in the browser. This is different from `performance.now()` which gives you the timestamp relative to `navigationStart`. The question is, what's actually needed by this code?

Reply:
unfortunately it looks like that particular piece of code isn't merged into master.  it should be this:

```javascript

	// If the browser supports performance.now(), swap that in for BOOMR.now
	try {
		var p = boomr.getPerformance();
		if (p &&
		    typeof p.now === "function" &&
		    /\[native code\]/.test(String(p.now)) &&		// #545 handle bogus performance.now from broken shims
		    p.timing &&
		    p.timing.navigationStart) {
			boomr.now = function() {
				return Math.round(p.now() + p.timing.navigationStart);
			};
		}
	}
	catch (ignore) {
		// empty
	}
```

Reply:
let me see how easy it is to merge this in right now

Reply:
Thanks! That's what I thought.

My understanding is that this code is trying to mark when data from the User Timing API was last accessed. By marking that time, it knows which marks/measures to _not_ send when subsequent beacons are sent.

Let's say you do the following:

1. Set performance mark "A"
1. Send a beacon, in which performance mark "A" is sent
1. Set performance mark "B"
1. Send a second beacon, which does not include "B"

The problem is that during the sending of the first beacon, the plugin sets a `from` value that will look like `1510690156237`. When the second beacon is sent, it compares that value with the `startTime` and `duration` times in each of the marks/measures. These values look like `
112665.99`. The mark/measures values will _always_ be less than `BOOMR.now()` and therefore the values are discarded.

The plugin presumably does this because it does not want to clear marks that have already been beaconed. Thus, it tries to determine which marks were added _after_ the previous beacon was sent.

Reply:
Sorry, looks like it's already in there but I couldn't figure out how to search:
https://github.com/SOASTA/boomerang/blob/master/boomerang.js#L1964-L1979

Reply:
@querymetrics can you look into this?

Reply:
Seeing this in Chrome 61:

<img width="229" alt="screen shot 2017-11-14 at 2 19 50 pm" src="https://user-images.githubusercontent.com/921795/32802799-f2d439ec-c946-11e7-91e9-40e94c75422b.png">

Reply:
right, `BOOMR.getPerformance()` just returns the `performance` object in a browser independent way.

Reply:
Ok...I think I'm finally understanding what's going on. `BOOMR.now()` _should_ be a high precise timestamp from the unix epoch. When `performance.now()` is available, that timestamp is added to `performance.timing.navigationStart` to get a more accurate timestamp than `Date.now()` provides.

If I'm understanding this correctly, then `BOOMR.now()` is the incorrect timestamp to use.

What if we did something like:

```
BOOMR.getPerformance.now() || BOOMR.now()
```

In the fallback case, we'd still have this problem, but browsers that support the User Timing API should be fine.

Reply:
that would throw an exception if the shim doesn't exist.  I think this would work:

```javascript
var p = BOOMR.getPerformance()

// HiRes Timestamp supported
if (p && p.now) {
    now = p.now();
}
// HiRes Timestamp not supported, but RT Plugin exists and has figured out navigationStart
else if (BOOMR.plugins.RT && BOOMR.plugins.RT.navigationStart) {
    now = BOOMR.now() - BOOMR.plugins.RT.navigationStart();
}
// NavigationTiming not supported, but either BOOMR.t_lstart or BOOMR.t_start are set
// which are when boomerang started and not exactly when the page started
else {
    now = BOOMR.now() - (BOOMR.t_lstart || BOOMR.t_start);
}
```

This will do everything possible to get the right timestamp relative to navigationStart.

Reply:
This is great. I didn't know many of these things existed. I'll integrate this suggestion and see it is works.

Would it be worth moving this code to the main script to have a canonical function for `performance.now()`-esque values?